File: src/main/scala/templemore/onx/version3/WinScanner.scala
package templemore.onx.version3

import Token._

/**
 * @author Chris Turner
 */
class WinScanner {

  def win_?(grid: Grid) = winningToken(grid) != null

  def winningToken(grid: Grid): Token = {
    for ( line <- grid.lines ) {
      var noughtCount = 0
      var crossCount = 0
      for ( token <- line.tokens ) {
        if (Token.Nought == token) noughtCount += 1
        else if (Token.Cross == token) crossCount += 1
      }
      if (noughtCount == 3) return Token.Nought
      if (crossCount == 3) return Token.Cross
    }
    return null
  }
}

File: src/main/java/templemore/onx/version2/WinScanner.java
package templemore.onx.version2;

/**
 * @author Chris Turner
 */
public class WinScanner {

    public boolean isWin(final Grid grid) {
        return null != winningToken(grid);
    }

    public Token winningToken(final Grid grid) {
        for ( final Line line : grid.getAllLines() ) {
            int noughtCount = 0;
            int crossCount = 0;
            for ( final Token token : line.getTokens() ) {
                if ( Token.nought == token ) noughtCount++;
                else if ( Token.cross == token ) crossCount++;
            }

            if ( noughtCount == Grid.WIDTH ) return Token.nought;
            if ( crossCount == Grid.WIDTH ) return Token.cross;
        }
        return null;
    }
}
